File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/c85eec31c0814d744d203ea2ff3ef0fe.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define len(arr) (sizeof(arr) / sizeof(arr[0]))

typedef struct {
    bool have_mod;
    int value;
    int mod;
} Node;

typedef struct {
    int *S;
    int n;
} TreeSegment;

int magic(int a, int b) {
    return a + b;
}

int apply_mod(int value, int mod, int tl, int tr) {
    return value + mod;
}

int merge_mods(int old, int _new, int tl, int tr) {
    return _new;
}

bool __intersects(int tl, int tr, int ql, int qr) {
    return !(tr <= ql || qr <= tl);
}

int __get_value(int *S, int i, int tl, int tr) {
    if (!S[i * 2].have_mod) return S[i * 2].value;
    return apply_mod(S[i * 2].value, S[i * 2].mod, tl, tr);
}

void __recalc_value(int *S, int i, int tl, int tr) {
    if (tl + 1 != tr) {
        int mid = (tl + tr) / 2;
        S[i].value = magic(
            __get_value(S, i + 1, tl, mid),
            __get_value(S, i + (mid - tl) * 2, mid, tr)
        );
    }
}

void __add_mod(int *S, int i, int tl, int tr, int sub) {
    if (S[i * 2].have_mod) {
        S[i * 2].mod = merge_mods(S[i * 2].mod, sub, tl, tr);
    } else {
        S[i * 2].mod = sub;
        S[i * 2].have_mod = true;
    }
}

void __push(int *S, int i, int tl, int tr) {
    if (S[i * 2].have_mod) {
        S[i * 2].value = apply_mod(S[i * 2].value, S[i * 2].mod, tl, tr);
        if (tl + 1 != tr) {
            int mid = (tl + tr) / 2;
            __add_mod(S, i + 1, tl, mid, S[i * 2].mod);
            __add_mod(S, i + (mid - tl) * 2, mid, tr, S[i * 2].mod);
        }
        S[i * 2].have_mod = false;
    }
}

void init(int *S, int i, int tl, int tr, int *src, int n) {
    if (tl + 1 == tr) {
        S[i * 2].value = src[tl];
    } else {
        int mid = (tl + tr) / 2;
        init(S, i + 1, tl, mid, src, n);
        init(S, i + (mid - tl) * 2, mid, tr, src, n);
        __recalc_value(S, i, tl, tr);
    }
}

void tree_segment_init(TreeSegment *ts, int *src, int n) {
    ts->n = n;
    ts->S = (int *)malloc(sizeof(int) * (2 * n + 1));
    init(ts->S, 0, 0, n, src, n);
}

int __query(int *S, int i, int tl, int tr, int ql, int qr) {
    __push(S, i, tl, tr);
    if (ql <= tl && tr <= qr) {
        return S[i * 2].value;
    } else {
        int mid = (tl + tr) / 2;
        if (__intersects(tl, mid, ql, qr) && __intersects(mid, tr, ql, qr)) {
            return magic(
                __query(S, i + 1, tl, mid, ql, qr),
                __query(S, i + (mid - tl) * 2, mid, tr, ql, qr)
            );
        } else if (__intersects(tl, mid, ql, qr)) {
            return __query(S, i + 1, tl, mid, ql, qr);
        } else {
            return __query(S, i + (mid - tl) * 2, mid, tr, ql, qr);
        }
    }
}

int tree_segment_query(TreeSegment *ts, int ql, int qr) {
    return __query(ts->S, 0, 0, ts->n, ql, qr);
}

void __change(int *S, int i, int tl, int tr, int j, int _new) {
    __push(S, i, tl, tr);
    if (tl + 1 == tr) {
        S[i * 2].value = _new;
    } else {
        int m = (tl + tr) / 2;
        if (j < m) {
            __change(S, S[i * 2].l, tl, m, j, _new);
        } else {
            __change(S, S[i * 2].r, m, tr, j, _new);
        }
        __recalc_value(S, i, tl, tr);
    }
}

void tree_segment_change(TreeSegment *ts, int i, int value) {
    __change(ts->S, 0, 0, ts->n, i, value);
}

void __update(int *S, int i, int tl, int tr, int ql, int qr, int mod) {
    __push(S, i, tl, tr);
    if (ql <= tl && tr <= qr) {
        __add_mod(S, i, tl, tr, mod);
    } else {
        int mid = (tl + tr) / 2;
        if (__intersects(tl, mid, ql, qr)) {
            __update(S, i + 1, tl, mid, ql, qr, mod);
        }
        if (__intersects(mid, tr, ql, qr)) {
            __update(S, i + (mid - tl) * 2, mid, tr, ql, qr, mod);
        }
        __recalc_value(S, i, tl, tr);
    }
}

void tree_segment_update(TreeSegment *ts, int ql, int qr, int mod) {
    __update(ts->S, 0, 0, ts->n, ql, qr, mod);
}

typedef struct {
    int *S;
    int n;
} TreeSegmentResp;

void init_resp(int *S, int i, int tl, int tr, int *src, int n) {
    if (tl + 1 == tr) {
        S[i * 2].value = src[tl];
    } else {
        int mid = (tl + tr) / 2;
        init_resp(S, i + 1, tl, mid, src, n);
        init_resp(S, i + (mid - tl) * 2, mid, tr, src, n);
        __recalc_value(S, i, tl, tr);
    }
}

void tree_segment_resp_init(TreeSegmentResp *ts, int *src, int n) {
    ts->n = n;
    ts->S = (int *)malloc(sizeof(int) * (2 * n + 1));
    init_resp(ts->S, 0, 0, n, src, n);
}

int __query_resp(int *S, int i, int tl, int tr, int ql, int qr) {
    __push(S, i, tl, tr);
    if (ql <= tl && tr <= qr) {
        return S[i * 2].value;
    } else {
        int mid = (tl + tr) / 2;
        if (__intersects(tl, mid, ql, qr) && __intersects(mid, tr, ql, qr)) {
            return magic(
                __query_resp(S, i + 1, tl, mid, ql, qr),
                __query_resp(S, i + (mid - tl) * 2, mid, tr, ql, qr)
            );
        } else if (__intersects(tl, mid, ql, qr)) {
            return __query_resp(S, i + 1, tl, mid, ql, qr);
        } else {
            return __query_resp(S, i + (mid - tl) * 2, mid, tr, ql, qr);
        }
    }
}

int tree_segment_resp_query(TreeSegmentResp *ts, int ql, int qr) {
    return __query_resp(ts->S, 0, 0, ts->n, ql, qr);
}

void __change_resp(int *S, int i, int tl, int tr, int j, int _new) {
    __push(S, i, tl, tr);
    if (tl + 1 == tr) {
        S[i * 2].value = _new;
    } else {
        int m = (tl + tr) / 2;
        if (j < m) {
            __change_resp(S, S[i * 2].l, tl, m, j, _new);
        } else {
            __change_resp(S, S[i * 2].r, m, tr, j, _new);
        }
        __recalc_value(S, i, tl, tr);
    }
}

void tree_segment_resp_change(TreeSegmentResp *ts, int i, int value) {
    __change_resp(ts->S, 0, 0, ts->n, i, value);
}

void __update_resp(int *S, int i, int tl, int tr, int ql, int qr, int mod) {
    __push(S, i, tl, tr);
    if (ql <= tl && tr <= qr) {
        __add_mod(S, i, tl, tr, mod);
    } else {
        int mid = (tl + tr) / 2;
        if (__intersects(tl, mid, ql, qr)) {
            __update_resp(S, i + 1, tl, mid, ql, qr, mod);
        }
        if (__intersects(mid, tr, ql, qr)) {
            __update_resp(S, i + (mid - tl) * 2, mid, tr, ql, qr, mod);
        }
        __recalc_value(S, i, tl, tr);
    }
}

void tree_segment_resp_update(TreeSegmentResp *ts, int ql, int qr, int mod) {
    __update_resp(ts->S, 0, 0, ts->n, ql, qr, mod);
}

int find_right(int *S, int i, int tl, int tr, int L, int sm) {
    __push(S, i, tl, tr);
    if (tr <= L) return -1;
    if (tl + 1 == tr) {
        int _this = magic(sm, __get_value(S, i, tl, tr));
        if (_this * 2 >= sm) {
            return tl;
        } else {
            return -1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (L <= tl) {
            int _left = magic(sm, __get_value(S, i + 1, tl, mid));
            if (_left * 2 >= sm) {
                return find_right(S, i + 1, tl, mid, L, sm);
            } else {
                return find_right(S, i + (mid - tl) * 2, mid, tr, L, _left);
            }
        } else {
            int t = find_right(S, i + 1, tl, mid, L, sm);
            if (t != -1) return t;
            return find_right(S, i + (mid - tl) * 2, mid, tr, L, t);
        }
    }
}

int tree_segment_find_right(int *S, int L, int sm) {
    return find_right(S, 0, 0, len(S) / 2, L, sm);
}

int find_left(int *S, int i, int tl, int tr, int R, int sm) {
    __push(S, i, tl, tr);
    if (R < tl) return -1;
    if (tl + 1 == tr) {
        int _this = magic(sm, __get_value(S, i, tl, tr));
        if (_this * 2 >= sm) {
            return tl;
        } else {
            return -1;
        }
    } else {
        int mid = (tl + tr) / 2;
        if (tr <= R + 1) {
            int _right = magic(sm, __get_value(S, S[i * 2].r, mid, tr));
            if (_right * 2 >= sm) {
                return find_left(S, S[i * 2].r, mid, tr, R, sm);
            } else {
                return find_left(S, S[i * 2].l, tl, mid, R, _right);
            }
        } else {
            int t = find_left(S, S[i * 2].r, mid, tr, R, sm);
            if (t != -1) return t;
            return find_left(S, S[i * 2].l, tl, mid, R, t);
        }
    }
}

int tree_segment_find_left(int *S, int R, int sm) {
    return find_left(S, 0, 0, len(S) / 2, R, sm);
}

int floor_mod(int a, int b) {
    if (a % b == 0) return 0;
    if (a >= 0 && b >= 0) return a % b;
    if (a <= 0 && b <= 0) return a % b;
    return abs(b) - (abs(a) % abs(b));
}

int main(int argc, char *argv[]) {
    int n, q;
    scanf("%d %d", &n, &q);
    int *pos = (int *)malloc(sizeof(int) * n);
    int *w = (int *)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &pos[i]);
    }
    for (int i = 0; i < n; i++) {
        scanf("%d", &w[i]);
    }

    TreeSegment ts;
    tree_segment_init(&ts, w, n);

    int *subw = (int *)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) {
        subw[i] = (w[i] * (i - pos[i])) % MOD;
    }

    TreeSegmentResp resp;
    tree_segment_resp_init(&resp, subw, n);

    for (int i = 0; i < q; i++) {
        int x, y;
        scanf("%d %d", &x, &y);
        if (x < 0) {
            int ind = -x;
            int _new = y;
            ind--;
            tree_segment_update(&ts, ind, ind + 1, _new);
            tree_segment_resp_update(&resp, ind, ind + 1, (_new * (ind - pos[ind])) % MOD);
        } else {
            int l = x, r = y;
            l--;

            int sm = tree_segment_query(&ts, l, r);

            int bestL = tree_segment_find_right(ts.S, l, sm);

            int x = pos[bestL] - bestL + l;

            int left = 0;
            if (l < bestL) {
                left = tree_segment_resp_query(&resp, l, bestL) + tree_segment_query(&ts, l, bestL) * (x - l);
            }
            int right = 0;
            if (bestL < r) {
                right = tree_segment_resp_query(&resp, bestL, r) + tree_segment_query(&ts, bestL, r) * (x - l);
            }
            printf("%d\n", floor_mod(left - right, MOD));
        }
    }

    free(pos);
    free(w);
    free(ts.S);
    free(subw);
    free(resp.S);

    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Python/c85eec31c0814d744d203ea2ff3ef0fe.py
import sys

class TreeSegment:
    def __init__(self, n, default_val, magic, apply_mod, merge_mods, mod):
        self.n = n
        self.S = [self.Node() for _ in range(2 * n + 1)]
        self.init(0, 0, len(default_val), default_val)
        self.magic = magic
        self.apply_mod = apply_mod
        self.merge_mods = merge_mods

    class Node:
        def __init__(self):
            self.have_mod = False
            self.value = None
            self.mod = None

    def __intersects(self, tl, tr, ql, qr):
        return not (tr <= ql or qr <= tl)

    def __get_value(self, i, tl, tr):
        if not S[i].have_mod:
            return S[i].value
        return apply_mod(S[i].value, S[i].mod, tl, tr)

    def __recalc_value(self, i, tl, tr):
        if tl + 1 != tr:
            mid = (tl + tr) // 2
            S[i].value = magic(
                __get_value(i + 1, tl, mid),
                __get_value(i + (mid - tl) * 2, mid, tr)
            )

    def __add_mod(self, i, tl, tr, sub):
        if S[i].have_mod:
            S[i].mod = merge_mods(S[i].mod, sub, tl, tr)
        else:
            S[i].mod = sub
            S[i].have_mod = True

    def __push(self, i, tl, tr):
        if S[i].have_mod:
            S[i].value = apply_mod(S[i].value, S[i].mod, tl, tr)
            if tl + 1 != tr:
                mid = (tl + tr) // 2
                __add_mod(i + 1, tl, mid, S[i].mod)
                __add_mod(i + (mid - tl) * 2, mid, tr, S[i].mod)
            S[i].have_mod = False

    def init(self, i, tl, tr, src):
        if tl + 1 == tr:
            S[i].value = src[tl]
        else:
            mid = (tl + tr) // 2
            init(i + 1, tl, mid, src)
            init(i + (mid - tl) * 2, mid, tr, src)
            __recalc_value(i, tl, tr)

    def __query(self, i, tl, tr, ql, qr):
        __push(i, tl, tr)
        if ql <= tl and tr <= qr:
            return S[i].value
        else:
            mid = (tl + tr) // 2
            if __intersects(tl, mid, ql, qr) and __intersects(mid, tr, ql, qr):
                return magic(
                    __query(i + 1, tl, mid, ql, qr),
                    __query(i + (mid - tl) * 2, mid, tr, ql, qr)
                )
            elif __intersects(tl, mid, ql, qr):
                return __query(i + 1, tl, mid, ql, qr)
            else:
                return __query(i + (mid - tl) * 2, mid, tr, ql, qr)

    def __change(self, i, tl, tr, j, _new):
        __push(i, tl, tr)
        if tl + 1 == tr:
            S[i].value = _new
        else:
            m = (tl + tr) // 2
            if j < m:
                __change(S[i].l, tl, m, j, _new)
            else:
                __change(S[i].r, m, tr, j, _new)
            __recalc_value(i, tl, tr)

    def __update(self, i, tl, tr, ql, qr, mod):
        __push(i, tl, tr)
        if ql <= tl and tr <= qr:
            __add_mod(i, tl, tr, mod)
        else:
            mid = (tl + tr) // 2
            if __intersects(tl, mid, ql, qr):
                __update(i + 1, tl, mid, ql, qr, mod)
            if __intersects(mid, tr, ql, qr):
                __update(i + (mid - tl) * 2, mid, tr, ql, qr, mod)
            __recalc_value(i, tl, tr)

    def query(self, ql, qr):
        assert ql < qr
        return __query(0, 0, n, ql, qr)

    def change(self, i, value):
        assert 0 <= i and i < n
        __change(0, 0, n, i, value)

    def update(self, ql, qr, mod):
        __update(0, 0, n, ql, qr, mod)

    def __find_right(self, i, tl, tr, L, checker, pref):
        __push(i, tl, tr)
        if tr <= L:
            return (-1, pref)
        if tl + 1 == tr:
            _this = magic(pref, __get_value(i, tl, tr))
            if checker(_this):
                return (tl, _this)
            else:
                return (-1, _this)
        else:
            mid = (tl + tr) // 2
            if L <= tl:
                _left = magic(pref, __get_value(i + 1, tl, mid))
                if checker(_left):
                    return __find_right(i + 1, tl, mid, L, checker, pref)
                else:
                    return __find_right(i + (mid - tl) * 2, mid, tr, L, checker, _left)
            else:
                t = __find_right(i + 1, tl, mid, L, checker, pref)
                if t[0] != -1:
                    return t
                return __find_right(i + (mid - tl) * 2, mid, tr, L, checker, t[1])

    def __find_left(self, i, tl, tr, R, checker, pref):
        __push(i, tl, tr)
        if R < tl:
            return (-1, pref)
        if tl + 1 == tr:
            _this = magic(pref, __get_value(i, tl, tr))
            if checker(_this):
                return (tl, _this)
            else:
                return (-1, _this)
        else:
            mid = (tl + tr) // 2
            if tr <= R + 1:
                _right = magic(pref, __get_value(S[i].r, mid, tr))
                if checker(_right):
                    return __find_left(S[i].r, mid, tr, R, checker, pref)
                else:
                    return __find_left(S[i].l, tl, mid, R, checker, _right)
            else:
                t = __find_left(S[i].r, mid, tr, R, checker, pref)
                if t[0] != -1:
                    return t
                return __find_left(S[i].l, tl, mid, R, checker, t[1])

    def init(self, src):
        n = len(src)
        self.S = [self.Node() for _ in range(2 * n + 1)]
        self.init(0, 0, len(src), src)

def floor_mod(a, b):
    if a % b == 0:
        return 0
    if a >= 0 and b >= 0:
        return a % b
    if a <= 0 and b <= 0:
        return a % b
    return abs(b) - (abs(a) % abs(b))

def magic(a, b):
    return a + b

def apply_mod(value, mod, tl, tr):
    return mod

def merge_mods(old, _new, tl, tr):
    return _new

class FastScanner:
    def __init__(self):
        self.input = sys.stdin.buffer
        self.buffer = bytearray()
        self.pos = 0

    def read(self, size):
        if self.pos >= len(self.buffer):
            self.buffer = self.input.read(size)
            self.pos = 0
        self.pos += size
        return self.buffer[self.pos - size:self.pos]

    def readline(self):
        res = bytearray()
        while True:
            b = self.read(1)
            if not b or b == b"\n":
                break
            res += b
        return res.decode()

    def read_int(self):
        s = self.readline().strip()
        return int(s)

    def read_ints(self):
        s = self.readline().strip()
        return list(map(int, s.split()))

    def read_string(self):
        return self.readline().strip()

def main():
    fs = FastScanner()
    n, q = fs.read_ints()
    pos = fs.read_ints()
    w = fs.read_ints()

    ts = TreeSegment(n, 0, magic, apply_mod, merge_mods, 1337)
    ts.init(w)

    subw = [0] * n
    for i in range(n):
        subw[i] = (w[i] * (i - pos[i])) % MOD

    resp = TreeSegment(n, 0, magic, apply_mod, merge_mods, 1337)
    resp.init(subw)

    for _ in range(q):
        x, y = fs.read_ints()
        if x < 0:
            ind = -x
            _new = y
            ind -= 1
            ts.update(ind, ind + 1, _new)
            resp.update(ind, ind + 1, (_new * (ind - pos[ind])) % MOD)
        else:
            l, r = x, y
            l -= 1

            sm = ts.query(l, r)

            bestL = ts.find_right(l, lambda x: x * 2 >= sm, 0)

            x = pos[bestL] - bestL + l

            left = 0
            if l < bestL:
                left = resp.query(l, bestL) + ts.query(l, bestL) * (x - l)
            right = 0
            if bestL < r:
                right = resp.query(bestL, r) + ts.query(bestL, r) * (x - l)

            print(floor_mod(left - right, MOD))

if __name__ == "__main__":
    main()
